feat: Add LRN opset 13 support

monicathieu opened the pull request:
Hi y'all,

After installing the package (through conda-forge! 🙏🏼 ), I went to convert my ONNX model, which is a modified AlexNet model (same layer architecture, just retrained with different image classes and new weights in the last layer). At this point, I realized that local response normalization was not yet implemented in onnx2torch! I looked at the relevant [pytorch docs](https://pytorch.org/docs/stable/generated/torch.nn.LocalResponseNorm.html?highlight=lrn) and [ONNX docs](http://www.xavierdupre.fr/app/mlprodict/helpsphinx/onnxops/onnx__LRN.html#id2), and it looks like both of them implement the formula directly from the [original AlexNet paper](https://proceedings.neurips.cc/paper/2012/file/c399862d3b9d6b76c8436e924a68c45b-Paper.pdf), just with slightly different argument names/order. Accordingly, I thought I might be able to handle adding a new converter for LRN.

I do want to note that I haven't added any tests for the converter, because that was beyond my Python/pytorch ability (this is all part of me learning!), but when I use my local version of the package to import my modified AlexNet model, it seeeeems not to have been mangled.

If you'd like me to work on adding the tests, I can, but it would take me some time to figure out the syntax. I would be thrilled if it ended up being very fast for you guys to add the tests yourselves.

senysenyseny16 commented on onnx2torch/node_converters/lrn.py:
```diff
@@ -0,0 +1,23 @@
+    size = node.attributes.get('size')
+    alpha = node.attributes.get('alpha', .0001)
+    beta = node.attributes.get('beta', .75)
+    k = node.attributes.get('size', 1)
```
`# pylint: disable=invalid-name`

senysenyseny16 commented on onnx2torch/node_converters/lrn.py:
```diff
@@ -0,0 +1,23 @@
+__all__ = []
```
Could you fix code style according to [pylint and black](https://github.com/ENOT-AutoDL/onnx2torch/runs/8181927682?check_suite_focus=true)?

senysenyseny16 commented:
Could you fix code style according to [pylint and black](https://github.com/ENOT-AutoDL/onnx2torch/runs/8181927682?check_suite_focus=true)?

senysenyseny16 commented:
> If you'd like me to work on adding the tests, I can, but it would take me some time to figure out the syntax. I would be thrilled if it ended up being very fast for you guys to add the tests yourselves.

It's not that hard, here's a rough plan:

1. Add `tests/node_converters/lrn_test.py`.
2. Create simple ONNX model consisting of single operation: `LRN`.
To do this you need to create ONNX node with the help of `onnx.helper.make_node` with `op_type='LRN'`, 4 attributes, input and output (https://github.com/onnx/onnx/blob/main/docs/Operators.md#LRN):

```python
def _test_lrn(data: np.ndarray, alpha: float, beta: float, bias: float, size: int) -> None:
    node = onnx.helper.make_node(
        op_type='LRN',
        inputs=list(test_inputs),
        outputs=['y'],
        alpha=alpha,  # ONNX attributes are passed as regular keyword arguments.
        beta=beta,
        bias=bias,
        size=size,
    )
```
3. Wrap your node to ONNX Model and call `check_onnx_model`
```python
    model = make_model_from_nodes(
        nodes=node,
        initializers={},
        inputs_example=test_inputs,
    )
    check_onnx_model(model, test_inputs)
```
`check_onnx_model` will convert ONNX LRN model to PyTorch and check that the ONNX and PyTorch models have the same outputs.
4. Parametrize `_test_lrn` function in `test_lrn` to cover LRN operator.

I think you can easily adapt [Transpose operation test](https://github.com/ENOT-AutoDL/onnx2torch/blob/main/tests/node_converters/transpose_test.py) for LRN.

monicathieu commented:
Ok, test is added! I modeled the LRN test after the transpose test, as well as the [batch norm test](https://github.com/ENOT-AutoDL/onnx2torch/blob/5424c8c78e7441552bd2057939f7cc20b997a4e1/tests/node_converters/batch_norm_test.py#L34). Batch normalization also takes some similar parameters to LRN, and I saw in that test those parameter values are drawn inside of `test_batch_norm()` using `np.random.uniform()`, so I used that implementation in `test_lrn()`.

senysenyseny16 commented on tests/node_converters/lrn_test.py:
```diff
@@ -0,0 +1,35 @@
+def test_lrn() -> None:  # pylint: disable=missing-function-docstring
+    shape = (227, 227, 3)
+    data = np.random.random_sample(shape).astype(np.float32)
+    alpha = np.random.uniform(low=0.0, high=1.0).astype(np.float32)
```
You don't need to convert `float` to `np.float32`, just rm `.astype(np.float32)`.
This is only needed for arrays (when you pass size to uniform).

```python
>>> import numpy as np
>>> alpha = np.random.uniform(low=0.0, high=1.0)
>>> type(alpha)
<class 'float'>
```

```python
>>> alpha = np.random.uniform(low=0.0, high=1.0).astype(np.float32)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'float' object has no attribute 'astype'
```

monicathieu replied:
thank you for clarifying! Will update.

senysenyseny16 commented on tests/node_converters/lrn_test.py:
```diff
@@ -0,0 +1,35 @@
+    alpha = np.random.uniform(low=0.0, high=1.0).astype(np.float32)
+    beta = np.random.uniform(low=0.0, high=1.0).astype(np.float32)
+    bias = np.random.uniform(low=1.0, high=5.0).astype(np.float32)
+    size = np.random.randint(low=1, high=10)
```
`size` should be odd, for example `n = size = 5`:

```
    center
      v
o x x c x x o o o o o (neurons)
  ---   ---
  n/2   n/2

o o x x c x x o o o o (next step)
o o o x x c x x o o o (next step)
o o o o x x c x x o o (next step)
...
```

To generate random odd you can use `randrange` from `random`:

```python
>>> import random
>>> random.randrange(start=1, stop=10, step=2)
9
>>> random.randrange(start=1, stop=10, step=2)
9
>>> random.randrange(start=1, stop=10, step=2)
1
>>> random.randrange(start=1, stop=10, step=2)
3
>>> random.randrange(start=1, stop=10, step=2)
1
>>> random.randrange(start=1, stop=10, step=2)
7
>>> random.randrange(start=1, stop=10, step=2)
9
>>> random.randrange(start=1, stop=10, step=2)
3
```

senysenyseny16 replied:
Also, to verify your test you can to install test dependencies:

```
pip install -r test-requirements.txt
```
and run pytest for your test:
```
pytest tests/node_converters/lrn_test.py
```

monicathieu replied:
Ah, thank you for clarifying that size needs to be odd and suggesting `random.randrange`, I'll use that. I had thought size acted as a radius and not a diameter, oops!

I just tested the test as well, and it works! with a caveat--right now, I'm testing on an M2 MacBook Pro that doesn't have NVIDIA GPUs, so CUDA isn't available for my machine. As such, I modified `tests/utils/common.py` to skip the cpu vs. CUDA check in `check_onnx_model()` if `torch.cuda.is_available()` returns False. (Not sure if I wrote it the most pythonically) Alas, I can't confirm that that test will pass right now 😅

senysenyseny16 replied:
This code will be more pythonic :)
```python
    _check_onnx_model(
        onnx_model=onnx_model,
        onnx_inputs=onnx_inputs,
        onnx_torch_check_function=onnx_torch_check_function,
        torch_cpu_cuda_check_function=torch_cpu_cuda_check_function if cuda_available else None,
        onnx_torch2onnx_check_function=onnx_torch2onnx_check_function,
        ignore_export_checker=ignore_export_checker,
        opset_version=opset_version,
    )
```

Could you revert "cuda_available" logic, because it skips all cuda-relative tests without any warnings/errors on CPU-only PC.
We will fix this problem later.

monicathieu replied:
ahhhh, I suspected there was a way to write the if-else inline, but I wasn't sure of the syntax!

I will go ahead and revert the change now (and just not run the tests anymore).

monicathieu commented on onnx2torch/node_converters/lrn.py:
```diff
@@ -0,0 +1,22 @@
+from onnx2torch.utils.common import OperationConverterResult
+from onnx2torch.utils.common import onnx_mapping_from_node
+
+
```
will do! I hadn't added this initially, but when I look again, I see that the only difference appears to be that ONNX LRN opset 1 doesn't allow tensors as type bfloat16 but LRN opset 13 does. Thus, this should be safe.

monicathieu commented on tests/node_converters/lrn_test.py:
```diff
@@ -0,0 +1,35 @@
+
+
+def test_lrn() -> None:  # pylint: disable=missing-function-docstring
+    shape = (227, 227, 3)
```
oof, I forgot about batch size and had the other stuff in the wrong order. Will update! thanks for catching.

monicathieu commented:
Just updated to add all your suggested changes--thanks for your help! I do want to flag that I changed `check_onnx_model()` not to run the cpu vs. CUDA check if `torch.cuda.is_available()` returns False, as I'm testing on a machine without NVIDIA GPUs. Please see if that code is to your taste, and I can modify if not.

senysenyseny16 commented on tests/node_converters/lrn_test.py:
```diff
@@ -0,0 +1,36 @@
+from random import randrange
```
https://github.com/ENOT-AutoDL/onnx2torch/runs/8241926516?check_suite_focus=true

senysenyseny16 replied:
To check your code before you committing it you can use [precommit hook](https://pre-commit.com/), we already set it up, just type `pre-commit install` in the root of onnx2torch directory.
This hook will check your changes when your type `git commit` and discard the commit if there is a problem.

monicathieu replied:
I installed pre-commit and isort and got the imports sorted! In order for pre-commit checks to pass, I had to rename my default branch on my fork (as it wouldn't allow pushing to branches called "main"--totally understand why but it was giving me issues on my fork!). Accordingly, I think that's caused this PR thread to die so I will try to reopen the PR using my renamed branch ... and then hopefully we'll be all set. Thank you for your patience!

senysenyseny16 replied:
It is not good practice even for forked repos:
1. there will be sync conflicts between main repository and your fork with commits on the top of main
2. to check new PR reviewer makes something like `git fetch origin pull/96/head:main ` that efficiently rewrites his main :)

Also, pre-commit only works on your side, after checking code you can disable it for pushing (`pre-commit uninstall`)

Anyway everything is fine :+1:

monicathieu replied:
Oooh! I wish GitHub would choose a different default branch name for forked repos then. Now I know for next time!

senysenyseny16 commented on tests/node_converters/lrn_test.py:
```diff
@@ -0,0 +1,36 @@
+import numpy as np
+import onnx
+
+from tests.utils.common import check_onnx_model  # pylint: disable=wrong-import-position
```
You don't need to use a `pylint: disable` here.

monicathieu replied:
It does seem to be fine now--previously when I was writing this, it started giving me wrong import position warnings, so I threw that in there to silence them. However, when I run linting now, that error isn't popping up. Huh! I'll remove the disable comments.
